### PR TITLE
DataSources: Azure Monitor, enable to query AppInsights()

### PR DESCRIPTION
**Pull request by BWalti**
**What this PR does / why we need it**:
The possibility to fetch available AppInsights, similar to (Log Analytics) "workspaces()", so that it hopefully can be used as a variable in "resource" in a panel query using "Azure Monitor Logs":
https://github.com/grafana/grafana/issues/34622#issuecomment-889780279

**Which issue(s) this PR fixes**:
Maybe it helps with #34622

**Special notes for your reviewer**:
I have no clue how to add the tests, I basically just "copy & pasted" the corresponding "workspaces()" code and adapted adequately. Sadly the build fails with the following (I have no experience in using Go):

``` console
Version: 8.1.0-pre, Linux Version: 8.1.0, Package Iteration: 1627984525pre
go build -ldflags -w -X main.version=8.1.0-pre -X main.commit=28a115cb96 -X main.buildstamp=1627984130 -X main.buildBranch=azure-monitor-query-appinsights -o ./bin/grafana-cli ./pkg/cmd/grafana-cli
build github.com/grafana/grafana/pkg/cmd/grafana-cli: cannot load embed: malformed module path "embed": missing dot in first path element
exit status 1
exit status 1
[Bra] 08-03 11:55:26 [ WARN] Fail to execute command: go [run build.go -dev build-cli] - exit status 1
```

Therefore I open this as a draft PR and hope to get some feedback :)

### CODE REVIEW

**Comment by joshhunt** [comment]
Thanks for the contribution @BWalti!

Just to understand the use case here - you want to use template variables to select an Application Insights resource when querying Azure Monitor Logs?

We have (soft) plans to move away from this janky style of template variables with the custom syntax (it's confusing to use, hard to maintain, and is overloaded with KQL queries you can also use in the same field), so we might not merge this PR if it will be deprecated shortly and/or add to maintenance burden, but I am interested in seeing if there's another solution to your problem without changing template variables in Azure Monitor.

**Comment by BWalti** [comment]
Hello @joshhunt 

I guess that was my intention, yes.
May I ask what your soft plan would be (as replacement?)? Because we are currently heavily relying on a lot of KQL queries in Panels using "Application Insights" and are now forced to move to "Logs", where I need to manually select the appropriate Azure Application Insights resource - disallowing us to create one dashboard for "multiple" environments (with multiple Application Insights instances).

What would be the intended way to go forward?

**Comment by joshhunt** [comment]
The end goal is to use Azure Resource Graph for template variables (and probably Azure Monitor Logs also), as that's designed for querying information about your Azure account itself.

I am curious now though - was there a scenario we made more difficult (or not possible at all) with the move from App Insights to Azure Monitor Logs? I understand that it's not yet possible to templatize other resource types for the Resource Picker in Logs, but our focus was on keep equivalent functionality for the App Insights -> Monitor Logs, and then adding in new (expected) functionality in later releases.

Was there a template variable or other functionality we broke or made not migratable in this change?

File: docs/sources/datasources/azuremonitor/template-variables.md [hunk 1]
```diff
@@ -14,21 +14,23 @@ types of template variables.
...
+| Name                                                                               | Description                                                                                                      |
+| ---------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------- |
+| `Subscriptions()`                                                                  | Returns subscriptions.                                                                                           |
+| `ResourceGroups()`                                                                 | Returns resource groups.                                                                                         |
+| `ResourceGroups(subscriptionID)`                                                   | Returns resource groups for a specified subscription.                                                            |
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/azuremonitor/template-variables.md:21]:
```suggestion
| `ResourceGroups(subscriptionID)`                                                   | Returns resource groups for the specified subscription.                                                            |
```

File: docs/sources/datasources/azuremonitor/template-variables.md [hunk 2]
```diff
@@ -14,21 +14,23 @@ types of template variables.
...
+| `MetricNames(aSubscriptionID, aMetricDefinition, aResourceName, aMetricNamespace)` | Returns a list of metric names for a specified subscription.                                                     |
+| `workspaces()`                                                                     | Returns a list of workspaces for the default subscription.                                                       |
+| `workspaces(subscriptionID)`                                                       | Returns a list of workspaces for the specified subscription (the parameter can be quoted or unquoted).           |
+| `AppInsights()`                                                                    | Returns a list of Application Insights for the default subscription.                                             |
+| `AppInsights(subscriptionID)`                                                      | Returns a list of Application Insights for the specified subscription (the parameter can be quoted or unquoted). |
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/azuremonitor/template-variables.md:33]:
```suggestion
| `AppInsights(subscriptionID)`                                                      | Returns a list of Application Insights for the specified subscription (the parameter can be within quotes or without). |
```

**Review by achatterjee-grafana — COMMENTED** [review]
Minor copy-edit suggestions.

**Comment by BWalti** [comment]
In my opinion the v8 upgrade broke the "simple" Azure Application Insights Query functionality (at least for us), meaning we could define a variable for the DataSource and one for the subscription, use these inside a Panel to define from which "Application Insights" we wanted to query and then just had to provide the corresponding KQL.
I have no clue however, what would have happened, if we had multiple Application Insight instances in this subscription (or even resource group) - but the credentials we used one specific for each environmental DataSource, only had access to one application insights per subscription.

So tl;dr: before v8, we could _switch_ the Application Insights instance _to query_ by using a simple _subscription variable_. Now we can't as we need to define the _resource_ on which this "Logs" thing should operate on. And there is no way of variable templating the Application Insights instance yet.

So this is basically rendering our "multi environment" dashboard useless, as I now had to _hard code_ one of the Application Insights instances. For the moment, the dashboard tries to use the prod instance, causing some panels to display errors in other environments (e.g. when switching the DataSource over to test environment).

**Comment by joshhunt** [comment]
@BWalti I hear you - it would be frustrating to lose this kind of functionality. It's not the intention of this rework - we want to make things easier to use, and more featureful. IMHO it doesn't make sense to have two different KQL interfaces when just one can query everything (ideally). 

I'm not quite sure if I follow how you were able to use template variables to switch between Application Insights resources - were you using a different data source (with different App Insights API keys?) for each environment, and using a data source environment variable to switch between them? I'm going to check out older versions to compare exactly what's happening. If you're able to, it would be a help if you could share your [dashboard JSON](https://grafana.com/docs/grafana/latest/dashboards/export-import/#exporting-a-dashboard) to see exactly how you've set it up. If you're able to do this, be sure to audit the JSON and replace anything sensitive.

In the meantime, I wonder if [Cross-resource queries](https://docs.microsoft.com/en-us/azure/azure-monitor/logs/cross-workspace-query) could be used to achieve what you're after?

Alternatively, if you have a fixed list of App Insights instances you want to query, you could create a manual template variable with a manual list you maintain. The values will need to be the full Azure Resource ID/URI of the App Insights resources you wish to query, which you could peak at least by viewing the JSON for a panel and it'll be under `logAnalytics.resource`. While not ideal, I think that could work.

**Comment by BWalti** [comment]
now as you mention it, right: I configured multiple Data Sources, each with it's own Application Insights API Key! Now I do understand how this could have worked technically :)

I may try to share a dashboard, but as you basically guessed completely right, I'm not sure if this is really needed anymore. If you still want to see one, I'd create a copy, throw out all "irrelevant" stuff and author sensitive data out of it (if anything would still be there). So would this still be helpful?

This fixed list of App Insights resources sounds promising! I guess I'll try that, thank you!

**Comment by BWalti** [comment]
Hey @joshhunt , sorry if I'm bothering you once again, but I have a quick follow up question:
Is there a way to "filter" the custom list of AAI instances using another variable value?

Meaning, if I have following variables:
- ENV = TST | QAE | PRD (type: Custom)
- DataSource = "Azure Monitor($ENV)" (type: DataSource)
- AAIs = TST : /subscription/.../...{more of these..}/microsoft.insights/components/{aai-instance-name}, QAE : ...... (type: Custom)

I would like to have the possibility to just toggle "ENV" (and therefore only showing that one on the dashboard) and all others get automatically "adapted" as there is only one possibility per each environment (I have similar variables for resourceGroupName using a "Regex" filter to achieve this).
So in Grafana wording we have "chained variables" (if I understand this correctly), where one variable influences the others.

I don't see a "filter" option on the "custom" variables (so the manually created list of Azure Application Instances), but maybe I can hack the json? there are "selected" attributes - maybe I could use this and use something like "$ENV == TST" ? Is this possible?
Or can I create some "logic variable" where the source would be my "AAI" variable (custom list) and I may filter there or something?

Kind regards and thank you very much for your help!

**Comment by joshhunt** [comment]
It doesn't look like there's the possibility to filter the custom list, which seems like a shame IMHO.

For better or worse, we could circle back to Azure for this. Instead of a _custom_ template variable, you could create a query template variable, use an Azure Monitor datasource (that has a default subscription, and at least one Log Analytics workspace available to it) with a query like the following:

```
datatable(env: string, uri: string)[
    "TST", "/sub/124/web/test-resource/",
    "PRD", "/sub/124/web/prod-resource/",
] | where env == "$environment" | project uri
```

It is quite the workaround, but it would use Kusto to construct our list and do the filtering.

**Comment by joshhunt** [comment]
Given we're moving to redo template variable queries for Azure Monitor in the medium term, we're not going to merge in something we would be deprecating pretty soon, so we'll be closing this PR. I understand this might be a bit disappointing, but you're still welcome to run your own branch if you wish. Either way, your contribution is appreciated!

If you have any feedback or feature requests, please feel free to open a discussion for it! https://github.com/grafana/grafana/discussions/categories/data-sources

**Pull request closed without merging**